testing/internal/e2e/tests/base/paginate_credential_test.go

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package base_test
+
+import (
+	"context"
+	"encoding/json"
+	"slices"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/hashicorp/boundary/api/credentials"
+	"github.com/hashicorp/boundary/api/scopes"
+	"github.com/hashicorp/boundary/testing/internal/e2e"
+	"github.com/hashicorp/boundary/testing/internal/e2e/boundary"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// TestCliPaginateCredentials asserts that the CLI automatically paginates to retrieve
+// all credentials in a single invocation.
+func TestCliPaginateCredentials(t *testing.T) {
+	e2e.MaybeSkipTest(t)
+	c, err := loadTestConfig()
+	require.NoError(t, err)
+
+	ctx := context.Background()
+	boundary.AuthenticateAdminCli(t, ctx)
+	newOrgId := boundary.CreateNewOrgCli(t, ctx)
+	t.Cleanup(func() {
+		ctx := context.Background()
+		boundary.AuthenticateAdminCli(t, ctx)
+		output := e2e.RunCommand(ctx, "boundary", e2e.WithArgs("scopes", "delete", "-id", newOrgId))
+		require.NoError(t, output.Err, string(output.Stderr))
+	})
+	newProjectId := boundary.CreateNewProjectCli(t, ctx, newOrgId)
+	newStoreId := boundary.CreateNewCredentialStoreStaticCli(t, ctx, newProjectId)
+
+	// Create enough credentials to overflow a single page.
+	client, err := boundary.NewApiClient()
+	require.NoError(t, err)
+	cClient := credentials.NewClient(client)
+	var credentialIds []string
+	for i := 0; i < c.MaxPageSize+1; i++ {
+		resp, err := cClient.Create(
+			ctx,
+			"username_password",
+			newStoreId,
+			credentials.WithUsernamePasswordCredentialUsername("user"),
+			credentials.WithUsernamePasswordCredentialPassword("password"),
+		)
+		require.NoError(t, err)
+		credentialIds = append(
+			credentialIds,
+			resp.Item.Id,
+		)
+	}
+
+	// List credentials
+	output := e2e.RunCommand(ctx, "boundary",
+		e2e.WithArgs(
+			"credentials", "list",
+			"-credential-store-id", newStoreId,
+			"-format=json",
+		),
+	)
+	require.NoError(t, output.Err, string(output.Stderr))
+
+	var initialCredentials credentials.CredentialListResult
+	err = json.Unmarshal(output.Stdout, &initialCredentials)
+	require.NoError(t, err)
+
+	var returnedIds []string
+	for _, cred := range initialCredentials.Items {
+		returnedIds = append(returnedIds, cred.Id)
+	}
+
+	require.Len(t, initialCredentials.Items, c.MaxPageSize+1)
+	assert.Empty(t, cmp.Diff(returnedIds, credentialIds, cmpopts.SortSlices(func(i, j string) bool { return i < j })))
```

Comment:
Do we want to add this to the targets test?

Reply:
Sure, why not :)